### PR TITLE
RunOnce Feeder. Should be persisted and controllable.

### PR DESCRIPTION
**Pull request by alexbergeron**
@mlaflamm @jgagnon1

### CODE REVIEW

File: mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+@RunWith(classOf[JUnitRunner])
+class TestTableOnceFeeder extends FunSuite {
+
+  /**
+   * Test feeder which use token as record. When loading records, it simply enumerates all the tokens in the
+   * specified token range.
+   */
+  class OnceTokenFeeder(val tokenRanges: TokenRangeSeq, limit: Int)
+    extends ResumableRecordDataFeeder with TableOnceFeeder {
```
Comment (mlaflamm) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala]: The implementation is very, very, very similar to `ContinuousTokenFeeder` defined in `TestTableContinuousFeeder`, only the traits mixed-in are different. I suggest to extract a common abstract class...

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    if (hasNext) false else {
+      context.data = context.data.updated("start-date", DateTime.now)
+      true
+    }
+  }
+
+  def reset(): Boolean = {
+    if (hasNext) false else {
+      lastRecord = None
+      currentRange = None
```
Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala:48]: I doubt this is very thread safe. This method is likely called from a thread different from the SPNL scheduler.

Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala:48]: This is also true for `start` and `restart`

Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala:48]: Who needs Thread safety? With all of this mutability, we're probably building a single-threaded application!

...Yeah, we'll have to do better than this. But I don't think making it `synchronized` would do it, since theses values might be modified from the main thread...

Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala:48]: I've got an idea using a volatile value or an atomic boolean... Can we discuss?

Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala:48]: Before the addition of these methods, the feeder was always running on the SPNL scheduler thread.

File: mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala (hunk 1 above)
Comment (alexbergeron) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala]: The `Feeder` in `TestTableContinuousFeeder` also has a `firstRangeLoadCount`, but it would just be unused in this case...

**Comment by alexbergeron** [comment]
Updated

File: mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+  test("feeder should not loop when loaded all data - large load limit") {
+    val ranges = Seq(TokenRange(2, 5), TokenRange(10, 12))
+    val feeder = new OnceTokenFeeder(ranges, limit = 1000)
+    feeder.init(TaskContext())
+    feeder.start()
+
+    val records = feeder.take(50).flatten.toList
+    records.flatMap(feeder.toRecord).take(15) should be(List(2, 3, 4, 5, 10, 11, 12))
+  }
```
Comment (jgagnon1) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala:94]: Not sure to see how you check that feeder is not looping on large limit ?

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableFeederByToken.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import com.wajam.spnl.TaskContext
+
+/**
+ * Feeder trait which ease creation of feeders who iterate over table records sequentially per range of tokens.
+ * When started it resume from the last acknowledged record saved in the task context.
+ */
+private[mysql] trait TableFeederByToken extends CachedDataFeeder with ResumableRecordDataFeeder with Logging {
+
+  private var currentContext: TaskContext = null
+  private[mysql] var lastRecord: Option[DataRecord] = None
```
Comment (jgagnon1) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableFeederByToken.scala:15]: since the class is already private[mysql] do you need it here to ?

Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableFeederByToken.scala:15]: Right, protected look like more appropriate in that case as it can only be accessed by subclasses.

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+  private def setDone(): Unit = {
+    updateContext(Map(finishedDate -> dateFormatter.print(DateTime.now)))
+  }
+
+  private def emptyResult =
+    (TokenRange(1, 0), lastRecord) // start is after end
+
+  private def updateContext(values: Map[String, String] = Map(), emptyContextToUpdate: Boolean = false): Map[String, String] = {
+    contextDataToUpdate synchronized {
```
Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala]: Using the mutable value we want to protect with a lock as the lock monitor is bad! You need to use a value that can never change during the lifetime of this feeder instance.

File: mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala (hunk 3 above)
Comment (alexbergeron) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala:94]: Maybe the name is not really appropriate...

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableFeederByToken.scala (hunk 4 above)
Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableFeederByToken.scala:15]: I think this might leak it in projects using it, extending a trait that extends this trait...

Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableFeederByToken.scala:15]: Ideally, I would only allow subclasses in the `mysql` package to access this.

Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableFeederByToken.scala:15]: After some check, `private[mysql]` does look like the better choice.

**Comment by alexbergeron** [comment]
Updated

**Comment by jgagnon1** [comment]
Much locking! I guess that is because it can be called from CRTX and SPNL right ? However, those will not be called that much I suppose so it's fine. If @mlaflamm suggested that to you, I'll be waiting for him to +1

**Comment by alexbergeron** [comment]
Yeah, basically, I'm locking on anything that might be executed on another thread than the job thread.. I'm trying to reduce locking to a minimum in the job.

File: mry-core/src/it/scala/com/wajam/mry/storage/mysql/FeederTestHelper.scala [hunk 6]
```diff
@@ -16,4 +17,34 @@ object FeederTestHelper {
...
+   * specified token range.
+   */
+  abstract class TokenFeeder(val tokenRanges: TokenRangeSeq, limit: Int)
+    extends ResumableRecordDataFeeder {
+
+    var firstRangeLoadCount = 0
+
+    type DataRecord = Long
+
+    val name = "ContinuousTokenFeeder"
```
Comment (mlaflamm) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/FeederTestHelper.scala]: `"TokenFeeder"`?

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/Job.scala [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.wajam.nrv.utils.Startable
+
+/**
+ * Marker trait allowing the definition of a Runnable Once job.
+ * Should be mixed in to a Feeder and to the SPNL Task running it
+ */
+trait Job {
+  def isStarted: Boolean
+  def getCurrentId: String
```
Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/Job.scala]: `def currentJobId: String`?

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        case None => emptyResult
+      }
+    } else emptyResult
+  }
+}
+
+object TableOnceFeeder {
+  private val startDate = "start-date"
+  private val finishedDate = "finished-date"
+  private val jobId = "job-id"
```
Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala]: `StartDate`, `FinishedDate` and `JobId`

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  def getCurrentId: String = Lock.synchronized(name + getJobId)
+
+  def isStarted: Boolean = hasNext
+
+  abstract override def toContextData(data: Feeder.FeederData): TaskContext.ContextData = super.toContextData(data) ++
+    Map(startDate -> getStartDate, finishedDate -> getFinishedDate, jobId -> getJobId)
+
+  private[mysql] def getLoadPosition: (TokenRange, Option[DataRecord]) = {
+    context.data ++= getContextToUpdate(emptyContextToUpdate = true)
+    if (hasNext) {
```
Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala]: It look like the whole method need to synchronized. What if `start` or `stop` are executed between these two lines?

```
  context.data ++= getContextToUpdate(emptyContextToUpdate = true)
  if (hasNext) {
```

Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala]: That might allow me to check less things...

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    (TokenRange(1, 0), lastRecord) // start is after end
+
+  private def updateContext(values: Map[String, String]): Unit = Lock.synchronized {
+    val updated = contextDataToUpdate ++ values
+    contextDataToUpdate = updated
+    updated
+  }
+
+
+  private def getContextToUpdate(emptyContextToUpdate: Boolean = false): Map[String, String] = Lock.synchronized {
```
Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala]: I must admit I don't get the purpose of `emptyContextToUpdate`. This method is invoked twice, once with the flag and once without. Wouldn't it be better to just in-line this method?

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+            setDone()
+            None
+          }
+          case None => None
+        }
+      }
+
+      loadRange match {
+        case Some(range) => (range, lastRecord)
+        case None if hasNext && !getContextToUpdate().contains(finishedDate) => (tokenRanges.head, None)
```
Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala]: Why `hasNext` is not sufficient and that a check of `finishedDate` is also necessary?

Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/TableOnceFeeder.scala]: Because `hasNext` only checks `context.data`, which hasn't been updaterd yet, since `setDone` affects only the temporary `contextDataToUpdate`

File: mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+import org.scalatest.junit.JUnitRunner
+import org.scalatest.FunSuite
+import org.scalatest.Matchers._
+import com.wajam.nrv.service.{TokenRangeSeq, TokenRange}
+import com.wajam.spnl.TaskContext
+import org.mockito.Mockito._
+import com.wajam.mry.storage.mysql.FeederTestHelper._
+
+@RunWith(classOf[JUnitRunner])
+class TestTableOnceFeeder extends FunSuite {
```
Comment (mlaflamm) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala:13]: Please add a test with an empty feeder to ensure it complete as expected.

Comment (alexbergeron) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala:13]: Done

File: mry-core/src/it/scala/com/wajam/mry/storage/mysql/FeederTestHelper.scala (hunk 6 above)
Comment (alexbergeron) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/FeederTestHelper.scala]: Forgotten

Comment (alexbergeron) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/FeederTestHelper.scala]: You won't like it, but I'll force a push with an updated commit

**Comment by alexbergeron** [comment]
Updated

File: mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    records2.flatMap(feeder.toRecord).take(15) should be(List())
+  }
+
+  test("should support empty feeders") {
+    val feeder = new OnceTokenFeeder(Nil, limit = 1)
+    feeder.init(TaskContext())
+    feeder.start("") should be(true)
+    val records = feeder.take(50).flatten.toList
+    records.flatMap(feeder.toRecord).take(15) should be(List())
+    feeder.stop() should be(true)
```
Comment (mlaflamm) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala]: I think you should verify `isStarted` before manually stopping it to ensure it stopped naturally beforehand.

**Comment by mlaflamm** [comment]
The `TableOnceFeeder` need to pass the JobId with every Record.

**Comment by alexbergeron** [comment]
> The TableOnceFeeder need to pass the JobId with every Record.

`TableOnceFeeder` can't define the appropriate method (`fromRecord`), even using `abstract override`, since it doesn't know about the actual type of `DataRecord`

**Comment by alexbergeron** [comment]
Disregard what I've just said an hour ago

**Comment by alexbergeron** [comment]
Updated

File: mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    records2.flatMap(feeder.toRecord).take(15) should be(List())
+  }
+
+  test("should support empty feeders") {
+    val feeder = new OnceTokenFeeder(Nil, limit = 1)
+    feeder.init(TaskContext())
+    feeder.start("") should be(true)
+    val records = feeder.take(50).flatten.toList
+    records.flatMap(feeder.toRecord).take(15) should be(List())
+    feeder.isStarted should be(true)
```
Comment (mlaflamm) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala]: Why true if the table is empty? After the table is completely iterated once, `isStarted` should be false. Right? Perhaps empty tables are special. I think it worth verifying `isStarted` in every tests...

Comment (mlaflamm) [inline mry-core/src/it/scala/com/wajam/mry/storage/mysql/TestTableOnceFeeder.scala]: I think that `TokenFeeder` is just not appropriate for this test.

**Comment by alexbergeron** [comment]
> I think that TokenFeeder is just not appropriate for this test.

Explain?

And if it's not appropriate, I waisted a sprint :crying_cat_face:

**Comment by mlaflamm** [comment]
I agree, your new F&O shirt doesn't fit you <b>waist</b>...

**Comment by alexbergeron** [comment]
Updated :crying_cat_face:

**Comment by mlaflamm** [comment]
+1
